x/auctions/keeper/msg_server.go

```diff
@@ -0,0 +1,88 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package keeper
+
+import (
+	"context"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	"github.com/evmos/evmos/v19/x/auctions/types"
+	"github.com/pkg/errors"
+)
+
+var _ types.MsgServer = &Keeper{}
+
+// Bid defines a method for placing a bid on an auction
+func (k Keeper) Bid(goCtx context.Context, bid *types.MsgBid) (*types.MsgBidResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	params := k.GetParams(ctx)
+	if !params.EnableAuction {
+		return nil, errorsmod.Wrapf(types.ErrAuctionDisabled, "auction is disabled")
+	}
+
+	lastBid := k.getHighestBid(ctx)
+	if bid.Amount.Amount.LTE(lastBid.Amount.Amount) {
+		return nil, errors.Wrapf(types.ErrBidMustBeHigherThanCurrent, "bid amount %s is less than last bid %s", bid.Amount, lastBid.Amount)
+	}
+
+	senderAddr, err := sdk.AccAddressFromBech32(bid.Sender)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid sender address")
+	}
+
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, senderAddr, types.ModuleName, sdk.NewCoins(bid.Amount)); err != nil {
+		return nil, errors.Wrap(err, "deposit failed")
+	}
+
+	// If there is a last bid, refund it
+	if !lastBid.Amount.IsZero() {
+		if err := k.refundLastBid(ctx); err != nil {
+			return nil, errors.Wrap(err, "refund failed")
+		}
+	}
+	k.setHighestBid(ctx, bid.Sender, bid.Amount)
+
+	// TODO: emit events
```

Comment:
should we address the events on this PR or in a following one?

Reply:
Still considering if we are going to just emit EVM events from the precompile instead of events on the Cosmos